caddyhttp/basicauth/basicauth.go

```diff
@@ -63,7 +63,7 @@ func (a BasicAuth) ServeHTTP(w http.ResponseWriter, r *http.Request) (int, error
 
 			// let upstream middleware (e.g. fastcgi and cgi) know about authenticated
 			// user; this replaces the request with a wrapped instance
-			r = r.WithContext(context.WithValue(r.Context(),
+			*r = *r.WithContext(context.WithValue(r.Context(),
```

Comment:
Ohhhh, is this because "the replacer" is "the replacer" that is being created by the logger in an earlier handler, and _that_ replacer is the one that holds a reference to the first `r`?

Yuck. Every time we do this it means we can't use the context for cancelling goroutines that may need it because it could be racey to pass it into a goroutine that accesses the context concurrently. (Maybe middlewares that need to use the context values can pull out what they need before the goroutine starts and pass them into the goroutine separately?)